File: lib/csharp/src/structure/TypeParameterConstraintClauseSyntax.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "csharp/inc/csharp.h"
#include "csharp/inc/Common.h"
#include "common/inc/WriteMessage.h"

#include "csharp/inc/messages.h"
#include <algorithm>
#include <string.h>
#include "common/inc/math/common.h"


namespace columbus { namespace csharp { namespace asg {

typedef boost::crc_32_type  Crc_type;

namespace structure { 
  TypeParameterConstraintClauseSyntax::TypeParameterConstraintClauseSyntax(NodeId _id, Factory *_factory) :
    Positioned(_id, _factory),
    ConstraintsContainer(),
    m_Name(0)
  {
  }

  TypeParameterConstraintClauseSyntax::~TypeParameterConstraintClauseSyntax() {
  }

  void TypeParameterConstraintClauseSyntax::prepareDelete(bool tryOnVirtualParent){
    while (!ConstraintsContainer.empty()) {
      const NodeId id = *ConstraintsContainer.begin();
      removeParentEdge(id);
      if (factory->getExistsReverseEdges())
        factory->reverseEdges->removeEdge(id, this->getId(), edkTypeParameterConstraintClauseSyntax_Constraints);
      ConstraintsContainer.pop_front();
    }
    removeName();
    base::Positioned::prepareDelete(false);
  }

  NodeKind TypeParameterConstraintClauseSyntax::getNodeKind() const {
    return ndkTypeParameterConstraintClauseSyntax;
  }

  ListIterator<structure::TypeParameterConstraintSyntax> TypeParameterConstraintClauseSyntax::getConstraintsListIteratorBegin() const {
    return ListIterator<structure::TypeParameterConstraintSyntax>(&ConstraintsContainer, factory, true);
  }

  ListIterator<structure::TypeParameterConstraintSyntax> TypeParameterConstraintClauseSyntax::getConstraintsListIteratorEnd() const {
    return ListIterator<structure::TypeParameterConstraintSyntax>(&ConstraintsContainer, factory, false);
  }

  bool TypeParameterConstraintClauseSyntax::getConstraintsIsEmpty() const {
    return getConstraintsListIteratorBegin() == getConstraintsListIteratorEnd();
  }

  unsigned int TypeParameterConstraintClauseSyntax::getConstraintsSize() const {
    unsigned int size = 0;
    ListIterator<structure::TypeParameterConstraintSyntax> endIt = getConstraintsListIteratorEnd();
    for (ListIterator<structure::TypeParameterConstraintSyntax> it = getConstraintsListIteratorBegin(); it != endIt; ++it) {
      ++size;
    }
    return size;
  }

  expression::IdentifierNameSyntax* TypeParameterConstraintClauseSyntax::getName() const {
    expression::IdentifierNameSyntax *_node = NULL;
    if (m_Name != 0)
      _node = dynamic_cast<expression::IdentifierNameSyntax*>(factory->getPointer(m_Name));
    if ( (_node == NULL) || factory->getIsFiltered(_node))
      return NULL;

    return _node;
  }

  bool TypeParameterConstraintClauseSyntax::setEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    switch (edgeKind) {
      case edkTypeParameterConstraintClauseSyntax_Constraints:
        addConstraints(edgeEnd);
        return true;
      case edkTypeParameterConstraintClauseSyntax_Name:
        setName(edgeEnd);
        return true;
      default:
        break;
    }
    if (base::Positioned::setEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  bool TypeParameterConstraintClauseSyntax::removeEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    switch (edgeKind) {
      case edkTypeParameterConstraintClauseSyntax_Constraints:
        removeConstraints(edgeEnd);
        return true;
      case edkTypeParameterConstraintClauseSyntax_Name:
        removeName();
        return true;
      default:
        break;
    }
    if (base::Positioned::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  void TypeParameterConstraintClauseSyntax::addConstraints(const structure::TypeParameterConstraintSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_NODE_IS_NULL);

    if (&(_node->getFactory()) != this->factory)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH);

    if (!(Common::getIsTypeParameterConstraintSyntax(*_node)))
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);

    ConstraintsContainer.push_back(_node->getId());
    setParentEdge(_node,edkTypeParameterConstraintClauseSyntax_Constraints);

    if (factory->reverseEdges)
      factory->reverseEdges->insertEdge(_node, this, edkTypeParameterConstraintClauseSyntax_Constraints);
  }

  void TypeParameterConstraintClauseSyntax::addConstraints(NodeId _id) {
    const structure::TypeParameterConstraintSyntax *node = dynamic_cast<structure::TypeParameterConstraintSyntax*>(factory->getPointer(_id));
    if (node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
    addConstraints( node );
  }

  void TypeParameterConstraintClauseSyntax::removeConstraints(NodeId id) {
    if (!factory->getExist(id))
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

    ListIterator<structure::TypeParameterConstraintSyntax>::Container::iterator it = find(ConstraintsContainer.begin(), ConstraintsContainer.end(), id);

    if (it == ConstraintsContainer.end())
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

    ConstraintsContainer.erase(it);

    removeParentEdge(id);

    if (factory->getExistsReverseEdges())
      factory->reverseEdges->removeEdge(id, this->getId(), edkTypeParameterConstraintClauseSyntax_Constraints);
  }

  void TypeParameterConstraintClauseSyntax::removeConstraints(structure::TypeParameterConstraintSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_EDGE_IS_NULL);

    removeConstraints(_node->getId());
  }

  void TypeParameterConstraintClauseSyntax::setName(NodeId _id) {
    expression::IdentifierNameSyntax *_node = NULL;
    if (_id) {
      if (!factory->getExist(_id))
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

      _node = dynamic_cast<expression::IdentifierNameSyntax*> (factory->getPointer(_id));
      if ( _node == NULL) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
      }
      if (&(_node->getFactory()) != this->factory)
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH );

      if (m_Name) {
        removeParentEdge(m_Name);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_Name, m_id, edkTypeParameterConstraintClauseSyntax_Name);
      }
      m_Name = _node->getId();
      if (m_Name != 0)
        setParentEdge(factory->getPointer(m_Name), edkTypeParameterConstraintClauseSyntax_Name);
      if (factory->getExistsReverseEdges())
        factory->reverseEdges->insertEdge(m_Name, this->getId(), edkTypeParameterConstraintClauseSyntax_Name);
    } else {
      if (m_Name) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);
      }
    }
  }

  void TypeParameterConstraintClauseSyntax::setName(expression::IdentifierNameSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);

    setName(_node->getId());
  }

  void TypeParameterConstraintClauseSyntax::removeName() {
      if (m_Name) {
        removeParentEdge(m_Name);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_Name, m_id, edkTypeParameterConstraintClauseSyntax_Name);
      }
      m_Name = 0;
  }

  void TypeParameterConstraintClauseSyntax::accept(Visitor &visitor) const {
    visitor.visit(*this);
  }

  void TypeParameterConstraintClauseSyntax::acceptEnd(Visitor &visitor) const {
    visitor.visitEnd(*this);
  }

  double TypeParameterConstraintClauseSyntax::getSimilarity(const base::Base& base){
    if(base.getNodeKind() == getNodeKind()) {
      return 1.0;
    } else {
      return 0.0;
    }
  }

  void TypeParameterConstraintClauseSyntax::swapStringTable(RefDistributorStrTable& newStrTable, std::map<Key,Key>& oldAndNewStrKeyMap ){
    std::map<Key,Key>::iterator foundKeyId;

      m_position.swapStringTable(newStrTable,oldAndNewStrKeyMap);

  }

  NodeHashType TypeParameterConstraintClauseSyntax::getHash(std::set<NodeId>& travNodes) const {
    if (hashOk) return nodeHashCache;
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_BEGIN,this->getId()); 
    if (travNodes.count(getId())>0) {
      common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_SKIP);
      return 0;
    }
    travNodes.insert(getId());
    Crc_type resultHash;
    resultHash.process_bytes( "structure::TypeParameterConstraintClauseSyntax", strlen("structure::TypeParameterConstraintClauseSyntax"));
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_END,resultHash.checksum()); 
    nodeHashCache = resultHash.checksum();
    hashOk = true;
    return nodeHashCache;
  }

  void TypeParameterConstraintClauseSyntax::save(io::BinaryIO &binIo,bool withVirtualBase  /*= true*/) const {
    Positioned::save(binIo,false);

    binIo.writeUInt4(m_Name);


    for (ListIterator<structure::TypeParameterConstraintSyntax>::Container::const_iterator it = ConstraintsContainer.begin(); it != ConstraintsContainer.end(); ++it) {
      binIo.writeUInt4(*it);
    }
    binIo.writeUInt4(0); // This is the end sign
  }

  void TypeParameterConstraintClauseSyntax::load(io::BinaryIO &binIo, bool withVirtualBase /*= true*/) {
    Positioned::load(binIo,false);

    m_Name =  binIo.readUInt4();
    if (m_Name != 0)
      setParentEdge(factory->getPointer(m_Name),edkTypeParameterConstraintClauseSyntax_Name);

    NodeId _id;

    _id = binIo.readUInt4();
    while (_id) {
      ConstraintsContainer.push_back(_id);
      setParentEdge(factory->getPointer(_id),edkTypeParameterConstraintClauseSyntax_Constraints);
      _id = binIo.readUInt4();
    }
  }


}


}}}

File: csharp/lib/csharp/Nodes/Structure/TypeParameterConstraintClauseSyntax.cs
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

using System;
using System.Text;
using System.Diagnostics;
using System.Collections.Generic;

namespace Columbus.Csharp.Asg.Nodes.Structure {

    /// <summary>
    /// TypeParameterConstraintClauseSyntax class, which represents the structure::TypeParameterConstraintClauseSyntax node.
    /// </summary>
    [DebuggerDisplay("{DemangledName}")]
#if (CSHARP_INTERNAL)
    internal
#else
    public
#endif
    class TypeParameterConstraintClauseSyntax : Columbus.Csharp.Asg.Nodes.Base.Positioned {


        // ---------- Edges ----------

        /// <summary>Container stores the id of the nodes the Constraints edge points to.</summary>
        protected LinkedList<uint> ConstraintsContainer;

        /// <summary>The id of the node the Name edge points to.</summary>
        protected uint m_Name;

        /// <summary>
        /// Constructor, only factory can instantiates nodes.
        /// </summary>
        /// <param name="nodeId">[in] The id of the node.</param>
        /// <param name="factory">[in] Poiter to the Factory the node belongs to.</param>
        public TypeParameterConstraintClauseSyntax(uint nodeId, Factory factory) : base(nodeId, factory) {
            ConstraintsContainer = new LinkedList<uint>();
            m_Name = 0;
        }

        /// <summary>
        /// Gives back the NodeKind of the node.
        /// </summary>
        /// <returns>Returns with the NodeKind.</returns>
        public override Types.NodeKind NodeKind {
            get { return Types.NodeKind.ndkTypeParameterConstraintClauseSyntax; }
        }


        // ---------- Reflection getter function ----------

        /// <summary>
        /// Gives back the begin iterator for the edge specified
        /// </summary>
        /// <param name="edge">[in] The multiple edge for which the iterator is requested</param>
        public override ListIterator<Columbus.Csharp.Asg.Nodes.Base.Base> GetListIteratorBegin(string edge) {
            switch(edge) {
                case "Constraints":
                    return ConstraintsListIteratorBegin.Convert<Columbus.Csharp.Asg.Nodes.Base.Base>();
                default:
                    return base.GetListIteratorBegin(edge);
            }
        }


        // ---------- Reflection add function ----------

        /// <summary>
        /// Adds item as a new edge
        /// </summary>
        public override void add(string edge, uint nodeId) {
            switch(edge) {
                case "Constraints":
                    addConstraints(nodeId);
                    return;
                default:
                    base.add(edge, nodeId);
                    return;
            }
        }


        // ---------- Edge getter function(s) ----------

        /// <summary>
        /// Gives back iterator for the Constraints edges.
        /// </summary>
        /// <returns>Returns an iterator for the Constraints edges.</returns>
        public ListIterator<Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintSyntax> ConstraintsListIteratorBegin {
            get {
                return new ListIterator<Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintSyntax>(ConstraintsContainer, fact, true);
            }
        }

        /// <summary>
        /// Gives back iterator for the Constraints edges.
        /// </summary>
        /// <returns>Returns an iterator for the Constraints edges.</returns>
        public ListIterator<Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintSyntax> ConstraintsListIteratorEnd {
            get {
                return new ListIterator<Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintSyntax>(ConstraintsContainer, fact, false);
            }
        }

        /// <summary>
        /// Tells whether the node has Constraints edges or not.
        /// </summary>
        /// <returns>Returns true if the node has at least one Constraints edge.</returns>
        public bool ConstraintsIsEmpty {
            get { return ConstraintsContainer.Count == 0; }
        }

        /// <summary>
        /// Gives back how many Constraints edges the node has.
        /// </summary>
        /// <returns>Returns with the number of Constraints edges.</returns>
        public uint ConstraintsSize {
            get { return (uint)(ConstraintsContainer.Count); }
        }

        /// <summary>
        /// Gives back the id of the node the Name edge points to.
        /// </summary>
        /// <returns>Returns the end point of the m_Name edge.</returns>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the edge id is invalid.</exception>
        public uint getName() {
            if (fact.getIsFiltered(m_Name))
                return 1;
            else
                return m_Name;
        }


        // ---------- Edge setter function(s) ----------

        /// <summary>
        /// Adds a new Constraints edge to the node and inserts it after the other ones.
        /// </summary>
        /// <param name="node">[in] The end point of the new Constraints edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if there is something wrong with the given node.</exception>
        public void addConstraints(Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintSyntax node) {
            addConstraints(node.Id);
        }

        /// <summary>
        /// Adds a new Constraints edge to the node and inserts it after the other ones.
        /// </summary>
        /// <param name="nodeId">[in] The end point of the new Constraints edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the given nodeId is invalid.</exception>
        public void addConstraints(uint nodeId) {
            if (!fact.getExist(nodeId))
                throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintClauseSyntax.addConstraints(NodeId)", "The end point of the edge does not exist.");
            Columbus.Csharp.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
            if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkTypeParameterConstraintSyntax)) {
                ConstraintsContainer.AddLast(nodeId);
            } else {
                throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintClauseSyntax.setConstraints(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
            }
           setParentEdge(nodeId, (uint)Types.EdgeKind.edkTypeParameterConstraintClauseSyntax_Constraints);
            if (fact.ReverseEdges != null)
              fact.ReverseEdges.insertEdge(nodeId, this.Id, Types.EdgeKind.edkTypeParameterConstraintClauseSyntax_Constraints);
        }

        /// <summary>
        /// Sets the Name edge.
        /// </summary>
        /// <param name="nodeId">[in] The new end point of the Name edge or 0. With 0 the edge can be deleted.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the nodeId is invalid.</exception>
        public void setName(uint nodeId) {
            if (nodeId != 0) {
                if (!fact.getExist(nodeId))
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintClauseSyntax.setName(NodeId)", "The end point of the edge does not exist.");

                Columbus.Csharp.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
                if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkIdentifierNameSyntax)) {
                    if (m_Name != 0) {
                        removeParentEdge(m_Name);
                    }
                    m_Name = nodeId;
                    setParentEdge(m_Name, (uint)Types.EdgeKind.edkTypeParameterConstraintClauseSyntax_Name);
                } else {
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.TypeParameterConstraintClauseSyntax.setName(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
                }
            } else {
                if (m_Name != 0) {
                    removeParentEdge(m_Name);
                }
                m_Name = 0;
            }
        }

        /// <summary>
        /// Sets the Name edge.
        /// </summary>
        /// <param name="node">[in] The new end point of the Name edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if there's something wrong with the given node.</exception>
        public void setName(Columbus.Csharp.Asg.Nodes.Expression.IdentifierNameSyntax node) {
            if (m_Name != 0) {
                removeParentEdge(m_Name);
            }
            m_Name = node.Id;
            setParentEdge(m_Name, (uint)Types.EdgeKind.edkTypeParameterConstraintClauseSyntax_Name);
        }


        // ---------- Accept functions for Visitor ----------

        /// <summary>
        /// It calls the appropriate visit method of the given visitor.
        /// </summary>
        /// <param name="visitor">[in] The used visitor.</param>
        public override void accept(Visitors.Visitor visitor) {
            visitor.visit(this);
        }

        /// <summary>
        /// It calls the appropriate visitEnd method of the given visitor.
        /// </summary>
        /// <param name="visitor">[in] The used visitor.</param>
        public override void acceptEnd(Visitors.Visitor visitor) {
            visitor.visitEnd(this);
        }

        /// <summary>
        /// Saves the node.
        /// </summary>
        /// <param name="io">[in] The node is written into io.</param>
        /// <exception cref="Columbus.ColumbusIOException">Throws ColumbusIOException if there is something wrong with the file.</exception>
        public override void save(IO io) {
            base.save(io);

            io.writeUInt4(m_Name);


            foreach (var Constraints in ConstraintsContainer) {
              io.writeUInt4(Constraints);
            }
            io.writeUInt4(0); // This is the end sign
        }

        /// <summary>
        /// Loads the node.
        /// </summary>
        /// <param name="io">[in] The node is read from io.</param>
        /// <exception cref="Columbus.ColumbusIOException">Throws ColumbusIOException if there is something wrong with the file.</exception>
        public override void load(IO binIo) {
            base.load(binIo);

            m_Name =  binIo.readUInt4();
            if (m_Name != 0)
              setParentEdge(m_Name,(uint)Types.EdgeKind.edkTypeParameterConstraintClauseSyntax_Name);

            uint _id;

            _id = binIo.readUInt4();
            while (_id != 0) {
              ConstraintsContainer.AddLast(_id);
              setParentEdge(_id, (uint)Types.EdgeKind.edkTypeParameterConstraintClauseSyntax_Constraints);
              _id = binIo.readUInt4();
            }
        }

    }

}
